packages/query-core/src/types.ts

```diff
@@ -223,6 +223,12 @@ export interface QueryOptions<
    * Maximum number of pages to store in the data of an infinite query.
    */
   maxPages?: number
+  /**
+   * If set to `true`, the query will warn if it is fetched on the server.
+   * Use this if you intend to only consume prefetched queries.
+   * Defaults to `false`.
+   */
+  warnOnServerFetches?: boolean
```

Comment:
bike shedding: discussed with co-workers, something like `strictPrefetch`/`strictPrefetches`? Happy with whatever the consensus is

Reply:
We could also consider different levels, so for example `strict` might lead to a runtime exception when compared to passing a `warn` value to simply log it 🤔 

> - given @TkDodo's [recent talk](https://youtu.be/QB-S6UU5YEU?si=RVi8kdV4jablIH4K&t=936), this should probably also accept a callback that gets the query(?) and not just a static boolean flag
> - warning log might be too specific, would it be better for this to be a function and allow the developer to control what should happen? Perhaps they want to log more/less info about the query to help track down the query that's missing a prefetch

I do like the callback approach more though, we could add an option called `onPrefetchMiss` or something along those lines.
This would give us more control over the outcome, e.g. could add observability, warning, throw an exception, etc.

Reply:
I do like the callback approach more though, we could add an option called onPrefetchMiss or something along those lines.
This would give us more control over the outcome, e.g. could add observability, warning, throw an exception, etc.

yea definetely see this useful for much more than just warning for un-prefetched queries. would love @TkDodo's thoughts on what API he prefers